### PR TITLE
2024 08 19/mahir/vendor update tags

### PR DESCRIPTION
**Pull request by Mahir12746**

### CODE REVIEW

**Review by Mobashir-Monim — CHANGES_REQUESTED** [review]
Please implement the `GetVendors` action making the same changes the the [account crud pr](https://github.com/Eveneer/evengoams-backend/pull/64).

**Comment by Mahir12746** [comment]
UpdateVendor Action

Phase 1

Inputs: new values as array in params and the vendor model instance
Output: change in the values of the vendor being updated
logic/process: the new values for a vendor will be replaced with the old values. Along with this if tags on the vendor changes then its relevant tags also needs to be synced with the vendor

**Comment by Mahir12746** [comment]
UpdateVendor Action

Phase 3

Line 1, PHP file has been opened.

Line 3, strict type checking enabled

Line 5, the namespace is App\Domains\Vendors\Actions, so to reach this file we have to go to App folder then Domains folder then Vendorss folder and finally to Actions folder where we will find the file

Line 7, we have injected dependency Vendor which is located in App\Domains\Vendors\Vendor
Line 8, we have injected VendorTypesEnum dependency from App\Domains\Vendors\Enums\VendorTypesEnum
Line 9, we have injected dependency Response from Illuminate\Auth\Access\Response
Line 10, we have injected dependency Request from Illuminate\Http\Request
Line 11, we have injected dependency ActionRequest from Lorisleiva\Actions\ActionRequest
Line 12, we have injected dependency AsAction from Lorisleiva\Actions\Concerns\AsAction

Line 14, the class name EditVendor which is also the file name

Line 16, we instructed the class to inherit the AsActions trait that was injected in line 12

Line 18, we have declared a function called authorize which accepts a HTTP request checking whether user has access or not. It returns the response model.

Line 20, it gets the user making the request

Line 22, the if condition checks if user has general access or not
Line 23, this returns response as true if the condition is true

Line 25, otherwise returns false with the message "You are unauthorised to perform this action"

Line 28, a function called handle is written taking vendor model as instance and an array called params. The return type of the function is vendor model instance

Line 31, if values for tag_ids given or not is checked
Line 32, if true then the tags associated with the vendor are synced

Line 35, the old values are replaced with the new values from params
Line 36, the vendor model instance is returned

Line 39, the rules function is written with return type being array

Line 42, the id is validated stating that it is required and must be present in the vendors table id column
Line 43, the name is validated stating that it is sometimes required and should be string along with the limit of characters that can be typed to 255
Line 44, the vendor type is checked from the VendorTypesEnum
Line 45, tag ids will be in an array and and sometimes required
Line 46, the tag ids exists in the tags table or not
Line 47, contacts will be in array and can be nulllable
Line 48, contact name will be string
Line 49, contact phone will be string
Line 50, email will be of email type

Line 54, ascontroller function is written with vendor model instance as input along with HTTP request

Line 56, the function returns the output of the handle function

Line 59, the jsonresponse function is taking vendor model instance and http request as input 

Line 62, the function is returning a message

**Comment by Mahir12746** [comment]
GetVendor Action

Phase 1

Inputs: id as string of the vendor we want to get
Outputs: the desired vendor or null if not found
Logic/Process: finding the vendor with the given id

**Comment by Mahir12746** [comment]
GetVendor Action

Phase 3

Line 1, the php file is opened

Line 3, declaration made that the file is strictly typed

Line 5, namespace is declared which says the file can be found by going to App folder then to Domains folder then Vendors and then to Actions and then finally to Queries

Line 7, dependency Vendor is in injected which can be found from App\Domains\Vendors\Vendor
Line 8, dependency response is injected from Illuminate\Auth\Access\Response
Line 9, dependency ActionRequest is injected from Lorisleiva\Actions\ActionRequest
Line 10, dependency AsAction is injected from Lorisleiva\Actions\Concerns\AsAction

Line 12, class name is written, GetVendor which is also the file name

Line 14, AsAction is inherited that we injected in line 10

Line 16, a function called authorize is written which is accepting HTTP request and the return type is Response

Line 18, the user variable stores the type of user

Line 20, this if statement checks whether the user has general access or not
Line 21, the if statement returns allowed if true

Line 24, otherwise the function returns denied with a message 'You are unauthorized to perform this action'

Line 27, handle function is written which is accepting id as string and the return type is Vendor model instance or null

Line 29, the function returns the Vendor which matches with the given id from the Vendor database table

Line 32, rules function is made which has no inputs but has output as an array

Line 35, the function validates the id by mentioning it to be required and that the id exists in the id column of the Vendors table

Line 39, asController function is made which accepts HTTP request

Line 41, returns the output of handle function which takes the id of the requested vendor from the request

Line 44, jsonResponse function is written which accepts the vendor model instance as the input parameter and the return type is array

Line 47, the function returns an array containing data of vendor
Line 48, and the success message

**Comment by Mahir12746** [comment]
GetVendors Action

Phase 1

Inputs: first input is the integer which represents the number of items to be shown per page and it is conditional so sometimes may not be given, its default value is 10. It also takes the search term as input
Outputs: the output can be all the vendors if the per page value not given or the number of vendors to be shown as output will be paginated by the per page value
Logic/Process: the logic is to get all the vendors if search term is not provided. If provided then based on the search term the vendors table will be searched to return any vendors if found.

**Comment by Mahir12746** [comment]
GetVendors Action

Phase 3

Line 1, php file is opened

Line 3, declaration made that the file is strictly typed

Line 5, namespace is written which tells the destination of the file

Line 7, Tag dependency is injected from App\Domains\Vendors\Vendor
Line 8, Response dependency is injected from Illuminate\Auth\Access\Response
Line 9, Collection dependency is injected from Illuminate\Database\Eloquent\Collection
Line 10, LengthAwarePaginator dependency is injected from Illuminate\Contracts\Pagination\LengthAwarePaginator
Line 11, ActionRequest dependency is injected from Lorisleiva\Actions\ActionRequest
Line 12, AsAction dependency is injected from Lorisleiva\Actions\Concerns\AsAction

Line 14, the name of the class is written which is also the name of the file

Line 16, AsAction is inherited from where it was injected in line 12

Line 18, a fuction called authorize is written which accepts a HTTP request as input and return type is Response instance

Line 20, user variable stores the data of the user making the request

Line 22, if condition checks whether user has general access or not
Line 23, the condition returns allow if has general access

Line 26, otherwise the function returns deny with a message

Line 29, a handle function is written which accepts an integer value for the number of items to be shown per page with default value 10 and a search term is being taken as input with default value empty string. The return type is Collection or LengthAwarePaginator

Line 31, Query builder is initiated for Vendor

Line 33, the if condition checks whether a search term was given or not
Line 35, the search term is being searched in the name column of vendor table
Line 36, the search term is being searched in the contacts column
Line 39, query is being returned based on a condition
Line 40, the get query gets all the vendors
Line 41, the query data is shown with a limit per page

Line 44, rules function is written whose return type is array

Line 47, the search term should be a string and nullable
Line 48, per_page value should be an integer, nullable and minimum one value should be there

Line 52, asController function accepts a HTTP request

Line 54, search_term variable takes the value as input
Line 55, per_page variable takes value as input

Line 57, the output of the handle function is being returned

Line 60, the jsonResponse function takes vendor model instance and HTTP request as input where return type is array

Line 62, message variable stores the count of vendors
Line 63, message variable stores found or fetched as string based on search term

Line 66, stores vendor model instance as value and data as key in an array
Line 67, a message is stored in a message variable

File: app/Domains/Vendors/Actions/CreateVendor.php [hunk 1]
```diff
@@ -41,7 +41,8 @@ public function rules(): array
     {
         return [
             'name' => ['required', 'string', 'max:255'],
-            'type' => ['required', 'in:' . implode(',', VendorTypesEnum::asArray())],
+            'type' => ['required',
```
Comment (Mobashir-Monim) [inline app/Domains/Vendors/Actions/CreateVendor.php]: Why did you separate the lines?

File: app/Domains/Vendors/Actions/Queries/GetVendors.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        return [
+            'search_term' => ['nullable', 'string'],
+            'per_page' => ['nullable', 'integer', 'min:1'],
+        ];
+    }
+
+    public function asController(ActionRequest $request)
+    {
+        $search_term = $request->input('search_term');
```
Comment (Mobashir-Monim) [inline app/Domains/Vendors/Actions/Queries/GetVendors.php]: Please read this [comment](https://github.com/Eveneer/evengoams-backend/pull/63/files#r1749168825)

File: app/Domains/Vendors/Actions/UpdateVendor.php [hunk 3]
```diff
@@ -27,6 +27,11 @@ public function authorize(ActionRequest $request): Response
 
     public function handle(Vendor $vendor, array $params): Vendor
     {
+
+        if (isset($params['tag_ids'])) {
+            $vendor->tags()->sync($params['tag_ids']);
+        }
+
         $vendor->update($params);
```
Comment (Mobashir-Monim) [inline app/Domains/Vendors/Actions/UpdateVendor.php:37]: You are going to have to unset `tag_ids` from param before you update.

**Review by Mobashir-Monim — CHANGES_REQUESTED** [review]

**Comment by Mahir12746** [comment]
GetVendor Action

Phase 1

Input: we pass the vendor id as parameters
Output: returns the vendor whose id matched from the vendors table
Logic/Process: find the vendor whose id matches with the given id

Phase 3

Line 1, opening of php code.

Line 3, declaration made that the file is strictly typed

Line 5, the namespace tells us how to access the file, which is inside the App folder in which there is Domains and in it there is Vendors and then Actions and finally Queries

Line 7, the Vendor model dependency is being injected,
Line 8, Response dependency is being injected from Illuminate\Auth\Access\Response.
Line 9, ActionRequest dependency is being injected from Lorisleiva\Actions\ActionRequest
Line 10, AsAction dependency is injected from  Lorisleiva\Actions\Concerns\AsAction.

Line 12, the class name is GetVendor which is also the file name.

Line 14, the AsAction trait was inherited by the class which was injected in lline 10.

Line 16, authorize function is written which is taking the request as parameter and its type is ActionRequest. The return type of the function is Response.

Line 18, from the request user data is being extracted and being stored in user variable.

Line 20, it is being checked whether in the extracted user data, there is general access or not.
Line 21, if yes then returning by executing the allow method from the Response class.

Line 24, otherwise from the Response class deny method is being called and returned.

Line 27, the handle function takes vendor id as parameter and the return type of the handle is either Vendor model instance or null.

Line 29, the function returns the specific vendor by finding the vendor using the provided id.

Line 32, the rules function has no parameter and the return type is array.

Line 35, id is being validated whether it exists in the vendors table or not.

Line 39, the asController function takes request as input with its type being ActionRequest.

Line 41, this function returns the output of the handle function with input id which is being extracted from request

Line 44, the jsonResponse function takes the vendor class object as parameter and the return type is array.

Line 47, a associative array is being returned with a message.

**Comment by Mahir12746** [comment]
GetVendors Action

Phase 1

Input: per_page and search_term per_page being integer and search_term being string
Output: returns a collection of all the data or paginated data is returned
Logic/Process: if search_term given then the vendors are searched accordingly otherwise all the vendors will be returned all together or by pagination.

Phase 3

Line 1, opening of php code.

Line 3, declaration made that the file is strictly typed

Line 5, the namespace tells us how to access the file, which is inside the App folder in which there is Domains and in it there is Vendors and then Actions and finally Queries

Line 7, the Vendor model dependency is being injected,
Line 8, Response dependency is being injected from Illuminate\Auth\Access\Response.
Line 9, Collection dependency is being injected from Illuminate\Database\Eloquent\Collection
Line 10, LengthAwarePaginator is being injected from Illuminate\Contracts\Pagination\LengthAwarePaginator
Line 11, ActionRequest is being injected from Lorisleiva\Actions\ActionRequest
Line 12, AsAction dependency is being injected from Lorisleiva\Actions\Concerns\AsAction

Line 14, the name of the class is given which is also the name of the file

Line 16, AsAction trait is being used here which was injected in line 12

Line 18, the authorize function accepts request object of ActionRequest and its return type is Response

Line 20, user variable stores the user data from the request

Line 22, if statement checks whether user has general access or not
Line 23, if true then allow method from Response is being executed and returned

Line 26, otherwise deny method from Response is being executed and returned

Line 29, handle function is written
Line 30, input per_page value can be integer or null and the default value is 10
Line 31 input search_term can be string or null and the default value is empty string
Line 32, the return type of the function is collection or lengthawarepaginator
Line 33, an empty query of the vendor is stored in query variable

Line 35, if statement checks whether search_term is given or not

Line 37, in the  query the search term is compared with the name to find similar named vendors
Line 38, or similar to contacts is searched

Line 41, will be returned after checking whether per_page contains null or not
Line 42, gets all the vendors using the query and returns it
Line 43, otherwise paginates the venodrs lists and then returns it

Line 46, function rules is written with return type being array

Line 49, search_term is validated by checking if it is string and can be nullable
Line 50, per_page is validated by checking if it is integer and can be nullable and also minimum number given should be 1

Line 54, asController function is wriiten with input parameter being request

Line 56, the function returns the handle function of this file with vendor and request as parameters

Line 59, jsonResponse function is written with parameters being vendor and request and the return type is array

Line 61, the message variable contains the count of vendors
Line 62, with the above message found or fetched is concatenated if search_term given

Line 65, the message is returned

**Comment by Mahir12746** [comment]
UpdateVendor Action

Phase 1

Input: the vendor to be updated is passed along with the parameters 
Output: the specific vendor updated with parameters
Logic/Process: to update the specific vendor with parameters and if tags are given for that vendor the tags are removed from the parameters bdefore updating the vendor, and those new tags are either created or the existing tags are synced with the vendor tags.


Phase 3

Line 1, opening of php code.

Line 3, declaration made that the file is strictly typed

Line 5, the namespace tells us how to access the file, which is inside the App folder in which there is Domains and in it there is Vendors and then Actions

Line 7, the CreateTags dependency is being injected from App\Domains\Tags\Actions\CreateTags
Line 8, the VendorTypesEnum dependency is being injected from App\Domains\Vendors\Enums\VendorTypesEnum
Line 9, Vendor model dependency is being injected from App\Domains\Vendors\Vendor
Line 10, Response dependency is being injected from Illuminate\Auth\Access\Response
Line 11, Request dependency is being injected from Illuminate\Http\Request
Line 12, ActionRequest dependency is being injected from Lorisleiva\Actions\ActionRequest
Line 13, AsAction is being injected from Lorisleiva\Actions\Concerns\AsAction

Line 15, the class name which is also the file name

Line 17, the AsAction trait is being used which has been injected in line 13

Line 19, authorize function is written which is taking request as input and return type is Response

Line 21, user variable stores the user data from request

Line 23, the if statement checks whether user has general excess or not 

Line 24, if yes then allow method of response is called and returned

Line 26, otherwise deny method of response is called and returned

Line 29, handle function is written, taking vendor and params as parameters with Vendor being the return type

Line 31, if tag_ids is given as input or not is checked

Line 32, if true then tag_ids param stores the values
Line 33, we unset the tag_ids from input parameter
Line 34, run method is called to execute the CreateTags action

Line 37, the vendor is updated with the param inputs
Line 38, the tags related to the vendor are synced
Line 39, the vendor is returned

Line 42, rules function is called

Line 45, id is being validated as required and must exist in vendor id column
Line 46, name is being validated as required sometimes and it being a string with maximum 255 character
Line 47, type is being validated to be only those that are present in VendorTypesEnum
Line 48, tag_ids field is being validated as required sometimes and it must be an array and can be nullable
Line 49, the input tags are validated so that they exists in id column of tags
Line 50, contacts is validated as being an array and can be nullable
Line 51, name within the contact is validated so that only string is given and can be nullable
Line 52, name within the phone is validated so that only string is given and can be nullable
Line 53, email within the contact is validated so that only email is given and can be nullable

Line 57, asController function is written with vendor and request as input
Line 59, the function returns the handle function with its inputs of this file

Line 62, jsonResponse function is written  taking vendor and request as parameter with array being the return type
Line 65, function returns an array which contains a message.

**Review by Mobashir-Monim — APPROVED** [review]

**Pull request merged**